lib/reversetunnel/emit_conn.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reversetunnel
+
+import (
+	"bytes"
+	"context"
+	"net"
+	"sync"
+
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/sshutils"
+)
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// emitConn is a wrapper for a net.Conn that emits an audit event for non-Teleport connections.
+type emitConn struct {
+	net.Conn
+	mu       sync.RWMutex
+	buffer   bytes.Buffer
+	emitter  apievents.Emitter
+	ctx      context.Context
+	serverID string
+	emitted  bool
+}
+
+func newEmitConn(ctx context.Context, conn net.Conn, emitter apievents.Emitter, serverID string) *emitConn {
+	return &emitConn{
+		Conn:     conn,
+		emitter:  emitter,
+		ctx:      ctx,
+		serverID: serverID,
+	}
+}
+
+func (conn *emitConn) Read(p []byte) (int, error) {
+	conn.mu.RLock()
+	n, err := conn.Conn.Read(p)
+
+	// Skip buffering if already could have emitted or will never emit.
+	if err != nil || conn.buffer.Len() == len(sshutils.ProxyHelloSignature) || conn.serverID == "" {
```

Comment:
That is much clearer - someone reading the code could follow ProxyHelloSignature around and figure out where it comes from. Thanks.